docs/spec/fastsync/scheduler.tla

```diff
@@ -0,0 +1,604 @@
+  /\ IF finishSc(outEvent') THEN
+       scRunning' = FALSE
+     ELSE UNCHANGED scRunning
+
+handleNoAdvanceExp ==
+    /\ inEvent.type = "tNoAdvanceExp"
+    /\ outEvent' = [type |-> "scFinishedEv", error |-> "timeout"]
+    /\ scRunning' = FALSE
+    /\ UNCHANGED <<scheduler>>
+
+NextSc ==
+  IF ~scRunning THEN
+    UNCHANGED <<outEvent, scRunning, scheduler>>
+  ELSE
+    \/ handleStatusResponse
+    \/ handleAddNewPeer
+    \/ handleRemovePeer
+    \/ handleTrySchedule
+    \/ handleBlockResponse
+    \/ handleNoBlockResponse
+    \/ handleBlockProcessed
+    \/ handleBlockProcessError
+    \/ handleNoAdvanceExp
+
+(* ----------------------------------------------------------------------------------------------*)
+(* The behavior of the environment.                                                              *)
+(* ----------------------------------------------------------------------------------------------*)
+
+InitEnv ==
+  /\ inEvent = noEvent
+  /\ envRunning = TRUE
+
+OnGlobalTimeoutTicker ==
+  /\ inEvent' = [type |-> "tNoAdvanceExp"]
+  /\ envRunning' = FALSE
+
+OnTrySchedule ==
+  /\ inEvent' = [type |-> "rTrySchedule"]
+  /\ UNCHANGED envRunning
+
+OnAddPeerEv ==
+  /\ inEvent' \in [type: {"bcAddNewPeer"}, peerID: PeerIDs]
+  /\ UNCHANGED envRunning
+
+OnStatusResponseEv ==
+  \* any status response can come from the blockchain, pick one non-deterministically
+  /\ inEvent' \in [type: {"bcStatusResponse"}, peerID: PeerIDs, height: Heights]
+  /\ UNCHANGED envRunning
+
+OnBlockResponseEv ==
+  \* any block response can come from the blockchain, pick one non-deterministically
+  /\ inEvent' \in [type: {"bcBlockResponse"}, peerID: PeerIDs, height: Heights, block: Blocks]
+  /\ UNCHANGED envRunning
+
+OnNoBlockResponseEv ==
+  \* any no block response can come from the blockchain, pick one non-deterministically
+  /\ inEvent' \in [type: {"bcNoBlockResponse"}, peerID: PeerIDs, height: Heights]
+  /\ UNCHANGED envRunning
+
+OnRemovePeerEv ==
+  \* although bcRemovePeer admits an arbitrary set, we produce just a singleton
+  /\ inEvent' \in [type: {"bcRemovePeer"}, peerID: PeerIDs]
+  /\ UNCHANGED envRunning
+
+OnPcBlockProcessed ==
+  /\ inEvent' \in [type: {"pcBlockProcessed"}, peerID: PeerIDs, height: Heights]
+  /\ UNCHANGED envRunning
+
+OnPcBlockVerificationFailure ==
+  /\ inEvent' \in [type: {"pcBlockVerificationFailure"}, firstPeerID: PeerIDs, secondPeerID: PeerIDs, height: Heights]
+  /\ UNCHANGED envRunning
+
+\* messages from scheduler
+OnScFinishedEv ==
+  /\ outEvent.type = "scFinishedEv"
+  /\ envRunning' = FALSE \* stop the env
+  /\ UNCHANGED inEvent
+
+NextEnv ==
+  IF ~envRunning THEN
+    UNCHANGED <<inEvent, envRunning>>
+  ELSE
+    \/ OnScFinishedEv
+    \/ OnGlobalTimeoutTicker
+    \/ OnAddPeerEv
+    \/ OnTrySchedule
+    \/ OnStatusResponseEv
+    \/ OnBlockResponseEv
+    \/ OnNoBlockResponseEv
+    \/ OnRemovePeerEv
+    \/ OnPcBlockProcessed
+    \/ OnPcBlockVerificationFailure
```

Comment:
We don't have pruning event. Do we want to add it for completness?

Reply:
Yes, i didn't add it in the end because the handler would just remove peers and this is currently also simulated by the environment that can send bcRemovePeer events at anytime.